File: AdventOfCode/2020/13-2.cpp
#include <iostream>
#include <vector>
#include <array>
#include <math.h>
#include <string>

using namespace std;

vector<array<int, 2>> get_IDs(string str)
{
    vector<string> str_temp;
    string temp = "";

    for (const auto& c : str )
    {
        if (c == ',')
        {
            str_temp.push_back(temp);
            temp = "";
        }
        else
        {
            temp += c;
        }
        
    }
    str_temp.push_back(temp);

    vector<array<int, 2>> end;
    for ( int i = 0; i < str_temp.size(); i++ )
    {
        if ( str_temp[i] != "x" )
        {
            end.push_back(array<int, 2>{i, stoi(str_temp[i])});
        }
    }
    return end;

}

bool valid(int offset, int id, int time)
{
    return (time + offset) % id == 0;
}

bool valid_for_all(vector<array<int, 2>> IDs, int time)
{
    for ( auto& bus : IDs )
    {
        if (!valid(bus[0], bus[1], time))
            return false;
    }
    return true;
}

auto get_max(vector<array<int, 2>> IDs)
{
   array<int, 2> end {0, 500000}; 

   for ( const auto& bus : IDs )
   {
       if ( bus[1] < end[1] )
       {
           end = bus;
       }
   }
   return end;
}

int looper(vector<array<int, 2>> IDs)
{
    auto bus = get_max(IDs);
    int time, index = 1;

    while (true)
    {
        time = index * bus[1] - bus[0];
        if ( valid_for_all(IDs, time) )
            return time;
        index++;
    }
}

int main(int argc, char const *argv[])
{
    cout << "Starting" << endl;
    auto IDs = get_IDs("29,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,37,x,x,x,x,x,409,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,17,13,19,x,x,x,23,x,x,x,x,x,x,x,353,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,41");
    cout << looper(IDs) << endl;
}

File: AdventOfCode/2020/13-2.py
def to_bus_id(bus_IDs):
    bus_IDs = bus_IDs.split(',')
    bus_IDs = [(i, int(el)) for i, el in enumerate(bus_IDs) if el != 'x']

    return bus_IDs

def valid(time, bus_IDs):
    for offset, bus in bus_IDs:
        if (time + offset) % bus != 0:
            return False
    return True

def time(bus_IDs):
    time = 100000000000000

    while True:
        print(time)
        if valid(time, bus_IDs):
            return time
        time += 1

def main():
    bus_id = to_bus_id('29,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,37,x,x,x,x,x,409,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,17,13,19,x,x,x,23,x,x,x,x,x,x,x,353,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,41')
    print(time(bus_id))

if __name__ == "__main__":
    main()